src/execution/flattenAsyncIterable.ts

```diff
@@ -0,0 +1,50 @@
+import { isAsyncIterable } from '../jsutils/isAsyncIterable';
+
+type AsyncIterableOrGenerator<T> =
+  | AsyncGenerator<T, void, void>
+  | AsyncIterable<T>;
+
+/**
+ * Given an AsyncIterable that could potentially yield other async iterators,
+ * flatten all yielded results into a single AsyncIterable
+ */
+export function flattenAsyncIterable<T, AT>(
+  iterable: AsyncIterableOrGenerator<T | AsyncIterableOrGenerator<AT>>,
+): AsyncGenerator<T | AT, void, void> {
+  const iteratorMethod = iterable[Symbol.asyncIterator];
+  const iterator: any = iteratorMethod.call(iterable);
+  let iteratorStack: Array<AsyncIterator<T>> = [iterator];
+
+  async function next(): Promise<IteratorResult<T | AT, void>> {
+    const currentIterator = iteratorStack[0];
+    if (!currentIterator) {
+      return { value: undefined, done: true };
+    }
+    const result = await currentIterator.next();
+    if (result.done) {
+      iteratorStack.shift();
+      return next();
+    } else if (isAsyncIterable(result.value)) {
+      const childIterator = result.value[
+        Symbol.asyncIterator
+      ]() as AsyncIterator<T>;
+      iteratorStack.unshift(childIterator);
```

Comment:
My assumption is that `unshift` is more likely to be `O(n)` than `push`. (I didn't bother to write my own benchmark but [this one I quickly found](https://www.measurethat.net/Benchmarks/Show/16089/0/pop-vs-shift-vs-unshift-vs-pop) certainly makes (in Chrome v8) push look much better than unshift.) Consider switching to push/pop.

Reply:
(although note that in #3703 I refactored things so that this function only needs to flatten precisely 2-level iterators with a single type that is always returned)

Reply:
(this is now gone)